Improved to be about 5x faster

gembleman opened the pull request:
| Benchmark          | Older Version | Newer Version(Use Libblur) | Newer Version(Implement directly) |
| ------------------ | ------------- | -------------------------- | --------------------------------- |
| **ssimulacra2**    | 20.16 ms      | 15.96 ms                   | 25.04ms                           |
| **blur**           | 11.61 ms      | 10.60 ms                   | 15.75ms                           |
| **downscale_by_2** | 1.94 ms       | 1.73 ms                    | 2.18ms                            |
| **image_multiply** | 2.35 ms       | 2.31 ms                    | 2.90ms                            |
| **ssim_map**       | 11.65 ms      | 4.43 ms                    | 4.63ms                            |
| **xyb_to_planar**  | 5.49 ms       | 5.12 ms                    | 7.52ms                            |



| Image Size    | Older Version | New Version(Use Libblur) | Performance Improvement | Score (Old version) | Score (New version / Use Libblur) | Accuracy Change |
| ------------- | ------------- | ------------------------ | ----------------------- | ------------------- | --------------------------------- | --------------- |
| **1448x1080** | 1.39s         | 264.70ms                 | 5.3x faster             | 17.392219           | 17.392059                         | -0.001%         |
| **2828x4242** | 11.02s        | 1.67s                    | 6.6x faster             | 84.031931           | 84.963643                         | -1.11%          |

Tested on Ryzen 3900x. Rayon feature has always been activated and measured.

I tried to improve the functions in lib.rs, but
most of them are improvements within the margin of error. Except for the ssim_map function.
The biggest improvement is Blur.
By using [libblur](https://github.com/awxkee/libblur) crate, the performance improved by at least 5x to at most 7x.
However, the evaluation score accuracy decreased by about 0.001% to 1.1%.
This seems like a tolerable tradeoff.



| Image Size    | Older Version | New Version(Implement directly) | Performance Improvement | Score (Old version) | Score (New version / Implement directly) | Accuracy Change |
| ------------- | ------------- | ------------------------------- | ----------------------- | ------------------- | ---------------------------------------- | --------------- |
| **1448x1080** | 1.39s         | 298.96ms                        | 4.5x faster             | 17.392219           | 17.385432                                | -0.01%          |
| **2828x4242** | 11.02s        | 2.69s                           | 4.0x faster             | 84.031931           | 84.052290                                | -0.02%          |

I've improved the implementation even when using a directly implemented blur function instead of libblur. 
The difference with the Older Version is that it transposes the input data, applies the horizontal filter, and then transposes it again.
I applied rayon to the vertical_pass as well. While accuracy decreases by 0.01%, it's about 4 times faster.

Edited. 2025-05-09

gembleman committed:
Version 0.6.0 / Up to 3x faster

gembleman committed:
add rustflags

gembleman committed:
add big size imgs for test

gembleman committed:
Run cargo check "--locked" delete

gembleman committed:
set Rust version 1.82.0
package `libblur v0.17.3` requires rustc 1.82.0 or newer

shssoichiro commented:
I'm pondering whether the decrease in accuracy is worth the speed difference. I'll ask some users to see what the general vibe is before moving forward. The blur has been found to be the biggest bottleneck, but as you also found, the only way that has been discovered to get significant performance gains is to emulate a gaussian blur instead of doing a true gaussian.

gembleman committed:
Added our own Gaussian filter implementation

gembleman committed:
Blur function improved. About 4x faster.

FreezyLemon commented on .github/workflows/rust.yml:
```diff
@@ -17,12 +17,12 @@ jobs:
       - name: Use predefined lockfile
         run: mv Cargo.lock.MSRV Cargo.lock
       - name: Build
-        run: cargo check --locked
```
This is here for a reason, if CI fails here then the lockfile is incorrect. Instead of removing `--locked`, update the `Cargo.lock.MSRV` file.

FreezyLemon reviewed (commented):
Thank you for the PR!

Looking at the [docs](https://docs.rs/libblur/latest/libblur/fn.gaussian_blur_f32.html) for libblur, it seems that [`EdgeMode::Clamp`](https://docs.rs/libblur/latest/libblur/enum.EdgeMode.html) means that for edges, anything outside the image is assumed to be == the edge pixel value. The gaussian_impl handles this differently IIRC, have you tried different settings to maybe improve the accuracy?

It should also be noted that the current gaussian implementation is already an approximation and not a true gaussian filter. So if anything, the libblur implementation should be compared to a real reference implementation. I would honestly not be surprised if the libblur implementation is both more performant and more accurate.

I like the idea of giving multiple blur algorithm options. (Maybe even at runtime, but that's out of scope for this PR).

Code-wise, the implementation looks good. I'd prefer the git history to be less messy but that probably won't matter for the squash-and-merge. CI should test all feature combinations if possible, if you're not familiar with GitHub actions I/we could just add this at the end.

gembleman committed:
libblur_impl accuracy improvement.

changed SIGMA from 2.3 to 2.2943

EdgeMode::Clamp, // diff 0.000 / 0.932

EdgeMode::Reflect101, // diff 0.008 / 0.933

EdgeMode::Reflect, // diff 0.003 / 0.934

EdgeMode::Constant, // diff 0.089 / 1.099

EdgeMode::Wrap, // diff 0.718 / 1.017

[cfg(not(feature = "rayon"))] added in blur_impl.rs

gembleman committed:
cleaned libblur_impl

gembleman committed:
Restore the 'expected' variable in the test_ssimulacra2() function back to its original state.

updated yuvxyb 0.4.1 -> 0.4.2

awxkee commented:
Just to clarify: `libblur::gaussian_blur` performs pure 2D separable convolution, as it should. The only factors that can affect the result are the system installed libm (used to generate filter weights, although you can invoke `libblur::filter1d` with your own weights if needed) and FMA availability. If you're working with extremely large intensities ( > 2e10 ) or extremely small ones ( < 2e-12 ), mandatory FMA or Dekker arithmetic may be required, which libblur does not provide.

If anything, if your current implementation is compiled by customers without explicitly setting `-Ctarget-feature=+fma` for a platform where LLVM could not detect FMA at compile time, performance might be interesting. Rust will emit software calls to `libm::fmaf`, and many math libraries do provide a software FMA fallback.

From signal theory alone, any Gaussian approximation makes the signal useless for any consequential analysis. After approximation your signal is composed from squares, triangles or whatever else your approximation uses. It makes no sense to perform any analysis on this.
